patches/api/0477-Add-paper-version-util-class.patch

```diff
@@ -0,0 +1,188 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: masmc05 <masmc05@gmail.com>
+Date: Fri, 16 Feb 2024 14:13:30 +0200
+Subject: [PATCH] Add paper version util class
+
+
+diff --git a/src/main/java/io/papermc/paper/util/ServerInfo.java b/src/main/java/io/papermc/paper/util/ServerInfo.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..931c6b704257b8baac7674c101d04ad7216d84e2
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/util/ServerInfo.java
+@@ -0,0 +1,98 @@
++package io.papermc.paper.util;
++
++import net.kyori.adventure.key.Key;
++import net.kyori.adventure.util.Services;
++import org.bukkit.UnsafeValues;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * A utility class to get information about the server
++ * Works even before Bukkit is initialized (e.g. on bootstrap)
++ */
++public class ServerInfo {
++    private ServerInfo() {
++        throw new UnsupportedOperationException("This class cannot be instantiated");
++    }
++    private static final ServerInfoProvider provider = Services.service(ServerInfoProvider.class).orElseThrow();
++
++    /**
++     * Get the version of the server
++     * @return the version of the server (e.g. "1.20.4", "1.20.2 Pre-release 2", "23w31a")
++     */
++    public static @NotNull String version() {
++        return provider.version();
++    }
++
++    /**
++     * Get the api version of the server
++     * @return the api version of the server (e.g. "1.20.4-R0.1-SNAPSHOT")
++     */
++    public static @NotNull String apiVersion() {
++        return provider.apiVersion();
++    }
++
++    /**
++     * Get the name of the server
++     * @return the name of the server (e.g. "Paper")
++     */
++    public static @NotNull String serverName() {
++        return provider.serverName();
++    }
++
++    /**
++     * Returns true if the server is a minecraft release version,
++     * false is it's a snapshot, pre-release, etc
++     * @return if the server is on a release version
++     */
++    public static boolean isRelease() {
++        return provider.isRelease();
++    }
++
++    /**
++     * Returns the unsafe values for the server for unsafe version values
++     * @return the unsafe values for the server
++     */
++    @Deprecated
++    @NotNull
++    public static UnsafeValues unsafe() {
```

Comment:
I think there is little point in exposing this.
Instead, expose protocol and data version on this (soon) interface.

Reply:
Alternatively we don't expose protocol version here but simply expose a static getting on the unsave instance or something. E.g. UnsaveValues.get().protocol version()

Exposing it here just feels a bit off, but I agree that I guess protocol and data version might be pretty magic.

Reply:
Will expose in a separate pr